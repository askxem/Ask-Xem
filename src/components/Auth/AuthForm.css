@import url('https://fonts.googleapis.com/css2?family=M+PLUS+Rounded+1c:wght@300&display=swap');
@import url('https://fonts.googleapis.com/css2?family=Varela+Round&display=swap');
@import url('https://fonts.googleapis.com/css2?family=Allerta+Stencil&display=swap');
/* font-family: 'Varela Round', sans-serif; */
/* font-family: 'M PLUS Rounded 1c', sans-serif; */
/* font-family: 'Allerta Stencil', sans-serif; */

.formcontainer {
  padding: 20px;
}

.formcontainer button {
  text-align: center;
  border: none;
  background-color: #91ad8c;
  border-radius: 20px;
  padding: 10px;
  box-shadow: 0px 5px 10px gray;
  color: white;
  font-family: 'Varela Round', sans-serif;
  font-size: 1em;
  cursor: pointer;
}

legend {
  font-family: 'Varela Round', sans-serif;
  margin-bottom: 10px;
}

fieldset,
form {
  border: none;
  display: flex;
  flex-direction: column;
  gap: 10px;
}

.formcontainer input {
  border: 1px solid grey;
  border-radius: 5px;
  height: 30px;
  padding: 10px;
}

.pwtoggle {
  text-align: right;
  font-size: 0.9em;
  cursor: pointer;
  text-decoration: underline;
}

@media only screen and (min-width: 768px) {
  .formcontainer {
    width: 50%;
    margin-left: 40px;
  }
}
