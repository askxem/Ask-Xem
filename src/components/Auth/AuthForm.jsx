import PropTypes from "prop-types"
import CoppaDisclaimer from "./CoppaDisclaimer.jsx"

/**
 * Called within the auth view/container component
 * @param {Boolean} isSigningUp Optional parameter, defeaults to false. Tells component how to manage the data. 
 * @returns a form that can handle signing up and logging in.
 */
export default function AuthForm({
  isSigningUp=false,
  email,
  password, 
  handleChange, 
  handleSubmit,
  isPasswordVisible,
  setIsPasswordVisible}) {

  return (
    <div>
      <fieldset>
        <legend>{isSigningUp ? 'Sign Up' : 'Login'}</legend>
        <form onSubmit={handleSubmit}>

          <label htmlFor="email">Email</label>
          <input 
          name='email' 
          type="email" 
          id="email" 
          value={email} 
          onChange={handleChange} 
          autoComplete="email" 
          required 
          />

          <label htmlFor="password">Password</label>
          <input 
          name='password'
          id="password"
          value={password}
          onChange={handleChange} 
          type={isPasswordVisible ? 'text' : 'password'} 
          autoComplete={isSigningUp ? 'new-password' : 'current-password'} 
          required 
          />
          
          {
            isSigningUp && <p aria-label='Password contraints'>{password.length >= 12 ? 'ðŸŸ¢ Password must be at least 12 characters long.' : 'ðŸ”´ Password must be at least 12 characters long.'}</p>
          }
          
          <div 
          id="toggle-password"
          aria-label={`Show password as plain text. Warning: this will display your password on the screen. Password is currently ${isPasswordVisible ? 'visible' : 'hidden'}`}
          onClick={() => setIsPasswordVisible(prevState => !prevState)}>
            { isPasswordVisible ? 'Hide Password' : 'Show Password'}
          </div>

          {
            isSigningUp
            ? (<>
              <CoppaDisclaimer />
              <button disabled={password.length < 12}>Create Account</button>
            </>)
            : <button>Login</button>
          }
          
        </form>
      </fieldset>


    </div>
  )
}

AuthForm.propTypes = {
  isSigningUp: PropTypes.bool,
  email: PropTypes.string,
  password: PropTypes.string,
  handleChange: PropTypes.func,
  handleSubmit: PropTypes.func,
  isPasswordVisible: PropTypes.bool,
  setIsPasswordVisible: PropTypes.func
}