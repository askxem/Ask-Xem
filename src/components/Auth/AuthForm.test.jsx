import { render } from '@testing-library/react';
import { Route, MemoryRouter } from 'react-router-dom';
import AuthForm from './AuthForm.jsx'

const testProps = {
    isSigningUp: false,
    email: 'test-email',
    password: 'test-password',
    handleChange: jest.fn(),
    handleSubmit: jest.fn(),
    isPasswordVisible: false,
    setIsPasswordVisible: jest.fn()
}

describe('renders authForm component', () => {
    
    it('renders login form', () => {
        const {container} = render(
        <MemoryRouter initialEntries={['/']}>
            <Route path='/'>
                <AuthForm {...testProps}/>
            </Route>
        </MemoryRouter>
        );
    
        expect(container).toMatchSnapshot();
    })

    it('renders signup form meeting password constraints', () => {
        const {container} = render(
            <MemoryRouter initialEntries={['/']}>
            <Route path='/'>
                <AuthForm {...testProps} isSigningUp={true}/>
            </Route>
        </MemoryRouter>
        );
    
        expect(container).toMatchSnapshot();
    })

    it('renders signup form failing password contraints', () => {
        const {container} = render(
            <MemoryRouter initialEntries={['/']}>
            <Route path='/'>
                <AuthForm {...testProps} password={'test'} isSigningUp={true}/>
            </Route>
        </MemoryRouter>
        );
    
        expect(container).toMatchSnapshot();
    })

})