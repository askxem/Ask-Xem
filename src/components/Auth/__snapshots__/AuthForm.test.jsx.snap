// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders authForm component renders login form 1`] = `
<div>
  <div
    class="formcontainer"
  >
    <fieldset>
      <legend>
        Login
      </legend>
      <form>
        <label
          for="email"
        >
          Email
        </label>
        <input
          autocomplete="email"
          id="email"
          name="email"
          required=""
          type="email"
          value="test-email"
        />
        <label
          for="password"
        >
          Password
        </label>
        <input
          autocomplete="current-password"
          id="password"
          name="password"
          required=""
          type="password"
          value="test-password"
        />
        <div
          aria-label="Show password as plain text. Warning: this will display your password on the screen. Password is currently hidden"
          class="pwtoggle"
          id="toggle-password"
        >
          Show Password
        </div>
        <button>
          Login
        </button>
        <a
          href="/signup"
        >
          Need an account?
        </a>
      </form>
    </fieldset>
  </div>
</div>
`;

exports[`renders authForm component renders signup form failing password contraints 1`] = `
<div>
  <div
    class="formcontainer"
  >
    <fieldset>
      <legend>
        Sign Up
      </legend>
      <form>
        <label
          for="email"
        >
          Email
        </label>
        <input
          autocomplete="email"
          id="email"
          name="email"
          required=""
          type="email"
          value="test-email"
        />
        <label
          for="password"
        >
          Password
        </label>
        <input
          autocomplete="new-password"
          id="password"
          name="password"
          required=""
          type="password"
          value="test"
        />
        <div
          aria-label="Show password as plain text. Warning: this will display your password on the screen. Password is currently hidden"
          class="pwtoggle"
          id="toggle-password"
        >
          Show Password
        </div>
        <p
          aria-label="Password contraints"
        >
          ðŸ”´ Password must be at least 12 characters long.
        </p>
        <div
          class="container"
        >
          <p>
            Disclaimer: You must be 13 years of age or older to create and manage an account.
            <a
              href="https://www.ftc.gov/enforcement/rules/rulemaking-regulatory-reform-proceedings/childrens-online-privacy-protection-rule"
              target="_blank"
            >
              <svg
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
                width="20px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8.228 9c.549-1.165 2.03-2 3.772-2 2.21 0 4 1.343 4 3 0 1.4-1.278 2.575-3.006 2.907-.542.104-.994.54-.994 1.093m0 3h.01M21 12a9 9 0 11-18 0 9 9 0 0118 0z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </a>
          </p>
        </div>
        <button
          disabled=""
        >
          Create Account
        </button>
        <a
          href="/login"
        >
          Already have an account?
        </a>
      </form>
    </fieldset>
  </div>
</div>
`;

exports[`renders authForm component renders signup form meeting password constraints 1`] = `
<div>
  <div
    class="formcontainer"
  >
    <fieldset>
      <legend>
        Sign Up
      </legend>
      <form>
        <label
          for="email"
        >
          Email
        </label>
        <input
          autocomplete="email"
          id="email"
          name="email"
          required=""
          type="email"
          value="test-email"
        />
        <label
          for="password"
        >
          Password
        </label>
        <input
          autocomplete="new-password"
          id="password"
          name="password"
          required=""
          type="password"
          value="test-password"
        />
        <div
          aria-label="Show password as plain text. Warning: this will display your password on the screen. Password is currently hidden"
          class="pwtoggle"
          id="toggle-password"
        >
          Show Password
        </div>
        <p
          aria-label="Password contraints"
        >
          ðŸŸ¢ Password must be at least 12 characters long.
        </p>
        <div
          class="container"
        >
          <p>
            Disclaimer: You must be 13 years of age or older to create and manage an account.
            <a
              href="https://www.ftc.gov/enforcement/rules/rulemaking-regulatory-reform-proceedings/childrens-online-privacy-protection-rule"
              target="_blank"
            >
              <svg
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
                width="20px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8.228 9c.549-1.165 2.03-2 3.772-2 2.21 0 4 1.343 4 3 0 1.4-1.278 2.575-3.006 2.907-.542.104-.994.54-.994 1.093m0 3h.01M21 12a9 9 0 11-18 0 9 9 0 0118 0z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </a>
          </p>
        </div>
        <button>
          Create Account
        </button>
        <a
          href="/login"
        >
          Already have an account?
        </a>
      </form>
    </fieldset>
  </div>
</div>
`;
